sdk/src/client/api/block_builder/transaction_builder/mod.rs

```diff
@@ -688,99 +690,117 @@ impl TransactionBuilder {
             }
         })
     }
+}
 
-    // Inputs need to be sorted before signing, because the reference unlock conditions can only reference a lower index
-    pub(crate) fn sort_input_signing_data(
-        mut inputs: Vec<InputSigningData>,
-        commitment_slot_index: SlotIndex,
-        committable_age_range: CommittableAgeRange,
-    ) -> Result<Vec<InputSigningData>, TransactionBuilderError> {
-        // initially sort by output to make it deterministic
-        // TODO: rethink this, we only need it deterministic for tests, for the protocol it doesn't matter, also there
-        // might be a more efficient way to do this
-        inputs.sort_by_key(|i| i.output.pack_to_vec());
-        // filter for ed25519 address first
-        let (mut sorted_inputs, account_nft_address_inputs): (Vec<InputSigningData>, Vec<InputSigningData>) =
-            inputs.into_iter().partition(|input_signing_data| {
-                let required_address = input_signing_data
-                    .output
-                    .required_address(commitment_slot_index, committable_age_range)
-                    // PANIC: safe to unwrap as non basic/account/foundry/nft outputs are already filtered out.
-                    .unwrap()
-                    .expect("expiration unlockable outputs already filtered out");
+#[derive(Clone, Debug, Default)]
+pub(crate) struct OrderedInputs {
+    ed25519: VecDeque<InputSigningData>,
+    other: BTreeMap<Address, VecDeque<InputSigningData>>,
+    len: usize,
+}
 
-                required_address.is_ed25519()
-            });
+impl OrderedInputs {
+    pub(crate) fn iter(&self) -> OrderedInputsIter<&InputSigningData> {
+        self.into_iter()
+    }
 
-        for input in account_nft_address_inputs {
-            let required_address = input
-                .output
-                .required_address(commitment_slot_index, committable_age_range)?
-                .expect("expiration unlockable outputs already filtered out");
+    pub(crate) fn insert(&mut self, input: InputSigningData, required_address: Address) {
+        if required_address.is_ed25519() {
```

Comment:
I think this need `is_ed25519_backed()` or `if let Some(ed25519) = required_address.backing_ed25519(){}` as they require the same Ed25519 signature

Reply:
okay